File: src/nl/weeaboo/kid/KIDUtil.java
package nl.weeaboo.kid;

import java.io.UnsupportedEncodingException;
import java.nio.ByteBuffer;
import java.util.List;

public final class KIDUtil {

   private KIDUtil() {
   }

   //Functions
   public static String readCString(ByteBuffer buf, int pos) {
      int oldlimit = buf.limit();
      int oldpos = buf.position();
      try {
         buf.limit(buf.capacity());
         buf.position(pos);
         return readCString(buf);
      } finally {
         buf.position(oldpos);
         buf.limit(oldlimit);
      }
   }

   /**
    * Reads a null-terminated UTF-8 string
    *
    * @param buf
    * @return
    */
   public static String readCString(ByteBuffer buf) {
      byte[] temp = new byte[256];
      int tempIndex = 0;

      while (buf.hasRemaining()) {
         byte b = buf.get();
         if (b == '\0') {
            break;
         }

         //Resize array if needed
         if (tempIndex >= temp.length) {
            byte[] newTemp = new byte[temp.length * 2];
            System.arraycopy(temp, 0, newTemp, 0, tempIndex);
            temp = newTemp;
         }
         temp[tempIndex++] = b;
      }

      return makeString(temp, 0, tempIndex);
   }

   public static String makeString(byte[] bytes) {
      return makeString(bytes, 0, bytes.length);
   }

   public static String makeString(byte[] bytes, int off, int len) {
      try {
         return new String(bytes, off, len, "MS932");
      } catch (UnsupportedEncodingException e) {
         return new String(bytes, off, len);
      }
   }

   public static String getLabel(int addr) {
      return String.format("L%08x", addr);
   }

   public static String getVarname(String v) {
      return "v_" + v;
   }

   public static String generateJumpScript(List<String> labels, List<String> targets,
           String labelPrefix, String returnAction, int itemsPerPage) {
      StringBuilder sb = new StringBuilder();

      int numPages = (targets.size() + itemsPerPage - 1) / itemsPerPage;
      for (int p = 0; p < numPages; p++) {
         StringBuilder lsb = new StringBuilder("\nchoice ");
         StringBuilder branches = new StringBuilder();

         int nstart = itemsPerPage * p;
         int n = nstart;
         for (int x = 0; x < itemsPerPage && n < targets.size(); x++) {
            if (x > 0) {
               lsb.append('|');
            }
            lsb.append(labels.get(n));

            branches.append("\nif selected == " + (n - nstart + 1));
            branches.append("\n    " + targets.get(n));
            branches.append("\nfi");

            n++;
         }

         if (numPages > 1) {
            n++;
            lsb.append("|[next]");
            branches.append("\nif selected == " + (n - nstart));
            branches.append("\n    goto " + labelPrefix + "p" + ((p + 1) % numPages));
            branches.append("\nfi");
         }

         n++;
         lsb.append("|[exit]");
         branches.append("\nif selected == " + (n - nstart));
         branches.append("\n    " + returnAction);
         branches.append("\nfi");

         sb.append("\nlabel " + labelPrefix + "p" + p);
         sb.append(lsb);
         sb.append(branches);
      }
      sb.append("\n");

      return sb.toString();
   }

}

File: tools/converter/KIDUtil.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace converter {
   internal static class KIDUtil {
      public static string ReadCString(BinaryReader buf, int pos) {
         long oldlimit = buf.BaseStream.Limit();
         long oldpos = buf.BaseStream.Position;
         try {
            buf.BaseStream.Limit(buf.BaseStream.Length);
            buf.BaseStream.Position = pos;
            return ReadCString(buf);
         }
         finally {
            buf.BaseStream.Position = oldpos;
            buf.BaseStream.Limit(oldlimit);
         }
      }
      public static string ReadCString(BinaryReader buf) {
         byte[] temp = new byte[256];
         int tempIndex = 0;

         while (buf.BaseStream.HasRemaining()) {
            byte b = buf.ReadByte();
            if (b == '\0') {
               break;
            }

            //Resize array if needed
            if (tempIndex >= temp.Length) {
               byte[] newTemp = new byte[temp.Length * 2];
               Array.Copy(temp, 0, newTemp, 0, tempIndex);
               temp = newTemp;
            }
            temp[tempIndex++] = b;
         }

         return MakeString(temp, 0, tempIndex);
      }
      public static string MakeString(byte[] bytes) {
         return MakeString(bytes, 0, bytes.Length);
      }
      public static string MakeString(byte[] bytes, int off, int len) {
         return Encoding.GetEncoding("shift_jis").GetString(bytes, off, len);
      }
      public static string GetVarname(string v) {
         return "v_" + v;
      }
   }
}
